src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java

```diff
@@ -413,115 +394,75 @@ private int toChars(long f, int e, FormattedFPDecimal fd) {
             y = t & MASK_28;
         }
-        lowDigits(l);
-        return NON_SPECIAL;
+        return lowDigits(str, index, l);
     }
 
-    private int toChars2(int h, int m, int l, int e) {
+    private int toChars2(byte[] str, int index, int h, int m, int l, int e) {
         /* -3 < e <= 0: plain format with leading zeroes */
-        appendDigit(0);
-        append('.');
+        putCharsAt(str, index, '0', '.');
+        index += 2;
         for (; e < 0; ++e) {
-            appendDigit(0);
+            putDigit(str, index++, 0);
         }
-        appendDigit(h);
-        append8Digits(m);
-        lowDigits(l);
-        return NON_SPECIAL;
+        putDigit(str, index, h);
+        put8Digits(str, index + 1, m);
+        return lowDigits(str, index + 9, l);
     }
 
-    private int toChars3(int h, int m, int l, int e) {
+    private int toChars3(byte[] str, int index, int h, int m, int l, int e) {
         /* -3 >= e | e > 7: computerized scientific notation */
-        appendDigit(h);
-        append('.');
-        append8Digits(m);
-        lowDigits(l);
-        exponent(e - 1);
-        return NON_SPECIAL;
+        putCharsAt(str, index, '0' + h, '.');
+        put8Digits(str, index + 2, m);
+        index = lowDigits(str, index + 10, l);
+        return exponent(str, index, e - 1);
     }
 
-    private void lowDigits(int l) {
+    private int lowDigits(byte[] str, int index, int l) {
         if (l != 0) {
-            append8Digits(l);
-        }
-        removeTrailingZeroes();
-    }
-
-    private void append8Digits(int m) {
-        /*
-         * Left-to-right digits extraction:
-         * algorithm 1 in [3], with b = 10, k = 8, n = 28.
-         */
-        int y = y(m);
-        for (int i = 0; i < 8; ++i) {
-            int t = 10 * y;
-            appendDigit(t >>> 28);
-            y = t & MASK_28;
-        }
-    }
-
-    private void removeTrailingZeroes() {
-        while (bytes[index] == '0') {
-            --index;
-        }
-        /* ... but do not remove the one directly to the right of '.' */
-        if (bytes[index] == '.') {
-            ++index;
+            put8Digits(str, index, l);
+            index += 8;
         }
+        return removeTrailingZeroes(str, index);
     }
 
-    private int y(int a) {
-        /*
-         * Algorithm 1 in [3] needs computation of
-         *     floor((a + 1) 2^n / b^k) - 1
-         * with a < 10^8, b = 10, k = 8, n = 28.
-         * Noting that
-         *     (a + 1) 2^n <= 10^8 2^28 < 10^17
-         * For n = 17, m = 8 the table in section 10 of [1] leads to:
-         */
-        return (int) (multiplyHigh(
-                (long) (a + 1) << 28,
-                193_428_131_138_340_668L) >>> 20) - 1;
-    }
-
-    private void exponent(int e) {
-        append('E');
+    private int exponent(byte[] str, int index, int e) {
+        putChar(str, index++, 'E');
         if (e < 0) {
-            append('-');
+            putChar(str, index++, '-');
```

Comment:
By the same logic as suggested above for `put8Digits`
```suggestion
            index = putChar(str, index, '-');
```

and other places as well.